lib/filter-schema.js

```diff
@@ -0,0 +1,166 @@
+
+module.exports = filterIntrospectionSchema
+
+function filterIntrospectionSchema (app, policy, { applyPolicy: policyFunction }) {
+  if (!app[kDirectiveGrouping]) {
+    app[kDirectiveGrouping] = []
+
+    // the filter hook must be the last one to be executed (after all the authContextHook ones)
+    app.ready(err => {
+      /* istanbul ignore next */
+      if (err) throw err
+      app.graphql.addHook('preExecution', filterGraphQLSchemaHook.bind(app))
+    })
+  }
+
+  app[kDirectiveGrouping].push({
+    policy,
+    policyFunction
+  })
+}
+
+filterIntrospectionSchema.updatePolicy = function (app, policy, { applyPolicy: policyFunction }) {
+  const storedPolicy = app[kDirectiveGrouping].find(({ policyFunction: storedPolicy }) => storedPolicy === policyFunction)
+  storedPolicy.policy = policy
+}
+
+async function filterGraphQLSchemaHook (schema, document, context) {
+  if (!isIntrospection(document)) {
+    return
+  }
+
+  const filteredSchema = await filterSchema(schema,
+    this[kDirectiveGrouping],
+    context)
+
+  return {
+    schema: filteredSchema
+  }
+}
+
+function isIntrospection (document) {
+  for (const queryType of document.definitions) {
+    if (queryType.operation !== 'query') {
+      // if there is a mutation or subscription, we can skip the introspection check
+      break
+    }
+
+    // if there is an introspection operation, we must filter the schema
+    if (queryType.selectionSet.selections.some(sel => (
+      sel.name.value === '__schema' ||
+      sel.name.value === '__type'
+    ))) {
+      return true
+    }
+  }
+  return false
+}
+
+async function filterSchema (graphQLSchema, policies, context) {
+  const filterDirectiveMap = {}
+
+  // each `policies` item is a directive
+  let skipFiltering = true
+  for (const { policy, policyFunction } of policies) {
+    // each `policy` contains all the GraphQL OBJECT and FIELDS that are affected by the directive
+    for (const [typeName, typePolicy] of Object.entries(policy)) {
+      // different `policies` item can affect the same GraphQL OBJECT
+      if (filterDirectiveMap[typeName] === undefined) {
+        filterDirectiveMap[typeName] = {}
+      } else if (filterDirectiveMap[typeName] === false) {
+        // if the object has already been filtered, we can skip the field processing
+        continue
+      }
+
+      const schemaType = graphQLSchema.getType(typeName)
+      const schemaTypeFields = typeof schemaType.getFields === 'function'
+        ? schemaType.getFields()
+        : {}
+      for (const [fieldName, fieldPolicy] of Object.entries(typePolicy)) {
+        // each `fieldName` is a single GraphQL item associated with the directive
+
+        if (filterDirectiveMap[typeName] === false || filterDirectiveMap[typeName][fieldName] === false) {
+          // if we have already decided to filter out this field
+          // it does not need to be checked again
+          continue
```

Comment:
I really like these comments you are making - they are super helpful - thanks! :)